### PR TITLE
Es6 refactoring

### PR DESCRIPTION
**Pull request by vvscode**

### CODE REVIEW

**Comment by ewnd9** [comment]
I usually use `proxyquire` like this:

```js
import Zip from 'jszip';

export default () => {
  const zip = new Zip();
  zip.file('collection.anki2', new Buffer('test'));

  return zip;
}
```
```js
// test.js
import test from 'ava';
import 'babel-register';

import proxyquire from 'proxyquire';
import sinon from 'sinon';

test(t => {
  const fileSpy = sinon.spy();
  
  function ZipMock() {
    this.file = fileSpy;
  }

  const fn = proxyquire.noCallThru().load('./index', {
    jszip: ZipMock
  }).default; // babel workaround

  fn();
  t.truthy(fileSpy.calledOnce);
  t.truthy(fileSpy.getCall(0).args[0] === 'collection.anki2');
  t.truthy(fileSpy.getCall(0).args[1] instanceof Buffer);
});
```

File: src/index.js [hunk 1]
```diff
@@ -48,11 +40,10 @@ export default function(deckName) {
...
 
-  update('update col set models=:models where id=1', { ':models': JSON.stringify(models) });
+  exporter.update('update col set models=:models where id=1', { ':models': JSON.stringify(models) });
 
-  return {
-    addMedia,
-    addCard,
-    save
-  };
+  return ['addCard', 'addMedia', 'save'].reduce((prev, i) => {
```
Comment (ewnd9) [inline src/index.js]: why is this all not in the `constructor` of  `exporter.js`?

Comment (vvscode) [inline src/index.js]: they are. We just  reexport not all `exporter` but only picked methods.

I'm not sure, you just don't export all methods before, and I try to not change external interface

**Comment by vvscode** [comment]
> I usually use proxyquire like this:

oh. That's instead of my `getZip` / `getDb` . I see. Ok, I'll use it

File: src/index.js (hunk 1 above)
Comment (ewnd9) [inline src/index.js]: No need for that, adding more functions doesn't break anything. You can apply it to anything which doesn't make sense to you. I will release it as a breaking change `v3.0.0` (as too much of the code are rewritten) when any new features are added.

Comment (vvscode) [inline src/index.js]: I suppose that these changes shouldn't be broken. But ok. I catch. Thx

Comment (ewnd9) [inline src/index.js]: As I showed in https://github.com/ewnd9/anki-apkg-export/pull/7#issuecomment-263780460 it's better to just return an instance of `Exporter` class

File: src/index.js [hunk 2]
```diff
@@ -11,24 +11,9 @@ export {SEPARATOR} from './exporter';
...
-      css: getCssTemplate(),
-      did: topDeckId,
-      id: topModelId,
-    });
-
-  return ['addCard', 'addMedia', 'save'].reduce((prev, i) => {
-    prev[i] = exporter[i];
-    return prev;
-  }, {});
+  return exporter
```
Comment (ewnd9) [inline src/index.js]: Why not like I suggested in https://github.com/ewnd9/anki-apkg-export/pull/7#issuecomment-263780460 ? `deckName` may still be passed as a first arguments (to prevent breaking changes) and the other options as second.

File: package.json [hunk 3]
```diff
@@ -6,6 +6,7 @@
   "dependencies": {
     "jszip": "^3.1.2",
     "sha1": "^1.1.1",
+    "sinon": "^1.17.6",
```
Comment (ewnd9) [inline package.json]: Should be in `devDependencies`

File: test/helpers.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
+import test from 'ava';
```
Comment (ewnd9) [inline test/helpers.js]: I've been thinking about this file and it seems unnecessary to me. Every function from `helpers` are used during the main tests, so it doesn't increase coverage.

File: test/exporter.js [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+import 'babel-polyfill';
+
+import Exporter from '../src/exporter';
+import { checksum } from '../src/helpers';
+
+let dbMock;
+let zipMock;
+let exporter;
+
+test.beforeEach(() => {
```
Comment (ewnd9) [inline test/exporter.js]: There is `t.context` which allows to pass variables from `beforeEach` to a test https://github.com/avajs/ava#before--after-hooks

`ava` runs tests concurrently so setting global (in terms of tests) variables are no-no

File: test/exporter.js [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    prepare: () => ({getAsObject: () => null })
+  };
+  zipMock = {
+    file: () => null,
+    generateAsync: () => null
+  };
+  exporter = new Exporter(dbMock, zipMock);
+});
+
+test('Exporter class exists', t => {
```
Comment (ewnd9) [inline test/exporter.js]: Same as https://github.com/ewnd9/anki-apkg-export/pull/8/files#r90220231 

`new Exporter !== new Exporter` It's JavaScript after all, so everything is possible. But, hypothetically, what is the point of overriding `new` behavior?

File: src/exporter.js [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+export default class {
+  constructor(db, zip) {
+    this.db = db;
+    this.zip = zip;
+    this.media = [];
+    this.topDeckId = rand();
+    this.topModelId = rand();
+    this.separator = SEPARATOR;
+
+    EXPORTED_METHODS.forEach(methodName => this[methodName] = this[methodName].bind(this));
```
Comment (ewnd9) [inline src/exporter.js]: Please remove it and `EXPORTED_METHODS`. There is no implication that they should be used without `this` in `master` branch (even though they can be used, it's a hidden from public surface)

File: src/helpers.js [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    sql = require('sql.js');
+  }
+  return sql;
+};
+
+export const rand = () => Math.random() * 100000000 | 0;
+
+export const getDb = () => new (getSql().Database)();
+
+export const getZip = (...args) => {
```
Comment (ewnd9) [inline src/helpers.js]: No need for that if you use `proxyquire` for testing

File: src/helpers.js [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+  const Zip = require('jszip');
+  return new Zip(...args);
+};
+
+export const getCssTemplate = () => {
```
Comment (ewnd9) [inline src/helpers.js]: Please return it as a multi-line variable

File: test/helpers.js (hunk 4 above)
Comment (vvscode) [inline test/helpers.js]: But in this case I create unit-test for each of them. And that make more easier to find what's wrong if main tests failed. 
If main test failed, but unit - not --- first should check code not covered with tests.

File: src/index.js (hunk 2 above)
Comment (vvscode) [inline src/index.js]: I'll do -- PR still in WIP ( I'll add mark about )

I just move code from one place to another without changing logic.

File: test/helpers.js (hunk 4 above)
Comment (ewnd9) [inline test/helpers.js]: I believe it's not a big enough project and not intended to ever be for this kind of thorough testing

> If main test failed, but unit - not --- first should check code not covered with tests.

It would make a sense if unit tests are separated and are ran firsts (no need to do it now)

File: src/index.js (hunk 2 above)
Comment (ewnd9) [inline src/index.js]: Ok, will wait when it becomes non wip before commenting

File: src/helpers.js (hunk 9 above)
Comment (vvscode) [inline src/helpers.js]: done

File: src/helpers.js (hunk 8 above)
Comment (vvscode) [inline src/helpers.js]: I still think that it be good if we able to provide dependencies ( db / zip ) from outside

It decrease coupling and make testing easier.

Yep. For this small project it's not critical, but it's pretty way

File: src/exporter.js (hunk 7 above)
Comment (vvscode) [inline src/exporter.js]: done

File: test/exporter.js (hunk 5 above)
Comment (vvscode) [inline test/exporter.js]: thx

File: package.json (hunk 3 above)
Comment (vvscode) [inline package.json]: done

File: src/index.js (hunk 2 above)
Comment (vvscode) [inline src/index.js]: done

File: src/helpers.js (hunk 8 above)
Comment (ewnd9) [inline src/helpers.js]: This is how all JavaScript modules I've ever encountered written (hard-code dependencies overridable by external tools like `proxyquire`). Perhaps, if you show some more repositories with such style of testing (every external dependency wrapped in a helper function and all of them passed in an entry point), I will be able to see the benefits.

Comment (ewnd9) [inline src/helpers.js]: Also, I would like to clarify the origin of all this. Is it because I said "Contributing lacking tests first is definitely a good choice" https://github.com/ewnd9/anki-apkg-export/pull/5#issuecomment-263160745 ? :no_mouth: I'm sorry, I meant it only in context of 4 commented lines at https://github.com/ewnd9/anki-apkg-export/commit/3c16fced9eb01185fc46bdf577626440af9dc959#diff-c1129c8b045390789fa8ff62f2c6b4a9L29

File: test/exporter.js [hunk 10]
```diff
@@ -82,7 +87,7 @@ test('Exporter.addCard', t => {
   t.is(notesUpdate[':sfld'], front);
   t.is(notesUpdate[':flds'], front + separator + back);
   t.is(notesUpdate[':mid'], topModelId);
-  t.is(notesUpdate[':csum'], checksum(front + separator + back));
+  t.is(notesUpdate[':csum'], exporter.checksum(front + separator + back));
```
Comment (ewnd9) [inline test/exporter.js]: `exporter.checksum('some string'), 2336613565` That style from above would be more appropriate here, instead of `exporter.checksum(front + separator + back)`, just the generated number

**Comment by vvscode** [comment]
@ewnd9  review, please

File: src/exporter.js [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+    this.db = this._getDb();
+    this.zip = this._getZip();
+    this.media = [];
+    this.topDeckId = rand();
+    this.topModelId = rand();
+    this.separator = SEPARATOR;
+    this.css = css;
+
+    return this
+      ._dbRun(this._getTemplate())
```
Comment (ewnd9) [inline src/exporter.js]: That 3 methods are used only here and they are private, just copy all the code from them to the constructor

File: src/exporter.js [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+      ':lapses': 0, // integer not null,
+      ':left': 0, // integer not null,
+      ':odue': 0, // integer not null,
+      ':odid': 0, // integer not null,
+      ':flags': 0, // integer not null,
+      ':data': '' // text not null
+    });
+  }
+
+  _checksum(str) { return parseInt(sha1(str).substr(0, 8), 16); }
```
Comment (ewnd9) [inline src/exporter.js]:
```js
_checksum(str) {
  return parseInt(sha1(str).substr(0, 8), 16);
}
```

File: src/exporter.js [hunk 13]
```diff
@@ -0,0 +1,174 @@
+const sha1 = require('sha1');
+
+export default class {
+  constructor(deckName) {
```
Comment (ewnd9) [inline src/exporter.js]: Add the second "options" argument as in https://github.com/ewnd9/anki-apkg-export/pull/7#issuecomment-263780460

```js
export default function(deckName) {
  return new ApkgExporter(deckName, { sql, template });
}
```

File: src/exporter.js [hunk 14]
```diff
@@ -0,0 +1,174 @@
+const sha1 = require('sha1');
```
Comment (ewnd9) [inline src/exporter.js]:
```js
import sha1 from 'sha1';
```

File: src/exporter.js [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+    if (process.env.APP_ENV === 'browser') {
+      require('script!sql.js');
+      sql = window.SQL;
+    } else {
+      sql = require('sql.js');
+    }
+    return new sql.Database();
+  }
+
+  _getZip(...args) {
```
Comment (ewnd9) [inline src/exporter.js]: Replace with following at the top 

```js
import Zip from 'jszip';
```

File: src/exporter.js [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+
+  _dbRun(...args) {
+    this.db.run(...args);
+    return this;
+  }
+}
+
+export const SEPARATOR = '\u001F';
+
+export const css = `.card {
```
Comment (ewnd9) [inline src/exporter.js]: Inline it in the constructor

File: test/exporter.js [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+test('Exporter._getTemplate', t => {
+  const { exporter } = t.context;
+
+  t.is(typeof exporter._getTemplate, 'function', 'should be a function');
+  let template = fs.readFileSync(path.join(__dirname, '../templates/template.sql'), 'utf-8');
+  t.is(exporter._getTemplate(), template, 'should return correct template');
+});
+
+
+test('Exporter._getZip', t => {
```
Comment (ewnd9) [inline test/exporter.js]: Remove this test

File: test/exporter.js [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+
+  t.is(typeof exporter._getZip, 'function', 'should be a function');
+
+  const zip = exporter._getZip();
+  t.truthy(typeof zip === 'object' && !!zip, 'should be an object');
+  t.is(typeof zip.file, 'function', 'zip should contains file method');
+  t.is(typeof zip.generateAsync, 'function', 'zip should contains generateAsync method');
+});
+
+test('rand', t => {
```
Comment (ewnd9) [inline test/exporter.js]: Remove this test

File: test/exporter.js [hunk 19]
```diff
@@ -0,0 +1,133 @@
...
+  t.is(typeof rand, 'function', 'should be a function');
+  t.is(typeof rand(), 'number', 'should return a number');
+});
+
+test('getLastItem', t => {
```
Comment (ewnd9) [inline test/exporter.js]: Remove this test

File: test/exporter.js [hunk 20]
```diff
@@ -0,0 +1,133 @@
...
+  t.is(notesUpdate[':mid'], topModelId);
+  t.is(notesUpdate[':csum'], exporter._checksum(front + separator + back));
+
+  t.is(exporterUpdateSpy.args[1][0],`insert into cards values(:id,:nid,:did,:ord,:mod,:usn,:type,:queue,:due,:ivl,:factor,:reps,:lapses,:left,:odue,:odid,:flags,:data)`);
+  const cardsUpdate = exporterUpdateSpy.args[1][1];
+  t.is(cardsUpdate[':did'], topDeckId);
+  t.is(cardsUpdate[':nid'], notesUpdate[':id'], 'should link both tables via the same note_id');
+});
+
+test('Exporter._getTemplate', t => {
```
Comment (ewnd9) [inline test/exporter.js]: Remove this test

File: src/exporter.js [hunk 21]
```diff
@@ -0,0 +1,174 @@
...
+    font-family: arial;
+    font-size: 20px;
+    text-align: center;
+    color: black;
+  }`;
+
+export const rand = () => Math.random() * 100000000 | 0;
+
+export const getLastItem = obj => {
```
Comment (ewnd9) [inline src/exporter.js]: Make it a private method in class

File: src/exporter.js [hunk 22]
```diff
@@ -0,0 +1,174 @@
...
+export const SEPARATOR = '\u001F';
+
+export const css = `.card {
+    font-family: arial;
+    font-size: 20px;
+    text-align: center;
+    color: black;
+  }`;
+
+export const rand = () => Math.random() * 100000000 | 0;
```
Comment (ewnd9) [inline src/exporter.js]: Make it a private method in class

File: test/exporter.js [hunk 23]
```diff
@@ -0,0 +1,133 @@
+import test from 'ava';
+import sinon from 'sinon';
+import fs from 'fs';
+import path from 'path';
+import proxyquire from 'proxyquire';
+
+import 'babel-register';
+import 'babel-polyfill';
+
+import { getLastItem, rand } from '../src/exporter';
```
Comment (ewnd9) [inline test/exporter.js]: Remove it

File: test/exporter.js [hunk 24]
```diff
@@ -0,0 +1,133 @@
...
+    this.file = () => null;
+    this.generateAsync = () => null;
+  },
+  sql: {
+    Database: function() {
+      this.export = () => "data";
+      this.prepare = () => ({getAsObject: () => null });
+      this.run = () => null;
+      this.exec = () => null;
+      }
```
Comment (ewnd9) [inline test/exporter.js]: indentation

File: test/exporter.js [hunk 25]
```diff
@@ -0,0 +1,133 @@
...
+    }
+}).default;
+
+test.beforeEach(t => {
+  t.context.exporter = new Exporter('testDeckName');
+});
+
+test('Exporter class exists', t => {
```
Comment (ewnd9) [inline test/exporter.js]: Remove this

File: test/exporter.js [hunk 26]
```diff
@@ -0,0 +1,133 @@
...
+  const { exporter } = t.context;
+
+  t.truthy(exporter instanceof Exporter, 'Exporter is constructor');
+});
+
+test('Exporter.addMedia', t => {
```
Comment (ewnd9) [inline test/exporter.js]: Remove this test

File: test/exporter.js [hunk 27]
```diff
@@ -0,0 +1,133 @@
...
+  t.truthy(dbExportSpy.called, 'should call .export on db');
+  t.truthy(zipFileSpy.calledWithMatch('collection.anki2'), 'should save notes/cards db');
+  t.truthy(zipFileSpy.calledWithMatch('media'), 'should save media');
+  t.truthy(zipFileSpy.calledWithMatch(0), 'should save media with two files');
+  t.truthy(zipFileSpy.calledWithMatch(1), 'should save media with two files');
+  t.truthy(zipGenerateAsyncSpy.called, 'should call zip.generateAsync');
+  t.truthy(['blob', 'nodebuffer'].includes(zipGenerateAsyncSpy.args[0][0].type), 'zip generates binary file');
+});
+
+test('Exporter._checksum', t => {
```
Comment (ewnd9) [inline test/exporter.js]: Remove this method

File: test/exporter.js [hunk 28]
```diff
@@ -0,0 +1,133 @@
...
+  const { exporter } = t.context;
+
+  t.is(typeof exporter._checksum, 'function', 'should be a function');
+  t.is(exporter._checksum('some string'), 2336613565, 'san calculate checksume for `some string`');
+});
+
+test('Exporter.addCard', t => {
+  const { exporter } = t.context;
+
+  t.is(typeof exporter.addCard, 'function', 'should be a function');
```
Comment (ewnd9) [inline test/exporter.js]: Remove this line

File: test/exporter.js [hunk 29]
```diff
@@ -0,0 +1,133 @@
...
+  const { topDeckId, topModelId, separator } = exporter;
+  const [front, back] = [5, 9, '!separator!', 'Test Front', 'Test back'];
```
Comment (ewnd9) [inline test/exporter.js]: What's up with 3 last items in array?

File: test/exporter.js [hunk 30]
```diff
@@ -0,0 +1,133 @@
...
+  exporter.addCard(front, back);
+
+  t.is(exporterUpdateSpy.callCount, 2, 'should made two requests');
+
+  t.is(exporterUpdateSpy.args[0][0], `insert into notes values(:id,:guid,:mid,:mod,:usn,:tags,:flds,:sfld,:csum,:flags,:data)`);
+  const notesUpdate = exporterUpdateSpy.args[0][1];
+  t.is(notesUpdate[':sfld'], front);
+  t.is(notesUpdate[':flds'], front + separator + back);
+  t.is(notesUpdate[':mid'], topModelId);
+  t.is(notesUpdate[':csum'], exporter._checksum(front + separator + back));
```
Comment (ewnd9) [inline test/exporter.js]: Replace with a generated checksum number

File: test/exporter.js [hunk 31]
```diff
@@ -0,0 +1,133 @@
...
+  t.deepEqual(exporter.media, [{ filename: 'some.file', data: 'data' }, { filename: 'another.file', data: 'new data' }]);
+});
+
+test('Exporter.save', t => {
+  const { exporter } = t.context;
+  const dbExportSpy = sinon.spy(exporter.db, 'export');
+  const zipFileSpy = sinon.spy(exporter.zip, 'file');
+  const zipGenerateAsyncSpy = sinon.spy(exporter.zip, 'generateAsync');
+
+  t.is(typeof exporter.save, 'function', 'should be a function');
```
Comment (ewnd9) [inline test/exporter.js]: Remove this line

File: src/exporter.js [hunk 32]
```diff
@@ -0,0 +1,174 @@
...
+      template = require('fs').readFileSync(__dirname + '/../templates/template.sql', 'utf-8');
+    }
+    return template;
+  }
+
+  _getFirstVal(query) {
+    return JSON.parse(this.db.exec(query)[0].values[0]);
+  }
+
+  _getDb() {
```
Comment (ewnd9) [inline src/exporter.js]: Move it to `index.js` and pass as a property of the second argument in constructor

File: src/exporter.js [hunk 33]
```diff
@@ -0,0 +1,174 @@
...
+    const { topDeckId } = this;
+    const decks = this._getInitialRowValue('col', 'decks');
+    const deck = getLastItem(decks);
+    deck.name = this.deckName;
+    deck.id = topDeckId;
+    decks[topDeckId + ''] = deck;
+    return this._update('update col set decks=:decks where id=1', { ':decks': JSON.stringify(decks) });
+  }
+
+  _getTemplate() {
```
Comment (ewnd9) [inline src/exporter.js]: Move it to `index.js` and pass as a property of the second argument in constructor

File: src/exporter.js (hunk 13 above)
Comment (vvscode) [inline src/exporter.js]: sql or db ? 

and I'm confused. Before you say to move dependencies into class. Now you say provide them on instantiating.

Comment (ewnd9) [inline src/exporter.js]: `sql`, because `require` is environment specific. 

Sorry for misdirection, I realized I was using the GitHub review feature incorrectly which led to a lot of scattered and unconnected comments. Now I am going to stick to a single review with "changes requested" status.

```js
// index.js
export default function(deckName) {
  return new ApkgExporter(deckName, { sql, template });
}
// exporter.js
constructor() {
  ...
  this.db = new sql.Database();
  ...
}
```

File: src/exporter.js [hunk 34]
```diff
@@ -0,0 +1,144 @@
+import sha1 from 'sha1';
+import Zip from 'jszip';
+
+
```
Comment (ewnd9) [inline src/exporter.js]: Remove unnecessary empty line

File: src/exporter.js [hunk 35]
```diff
@@ -0,0 +1,144 @@
...
+    this.topModelId = topModelId;
+    this.separator = SEPARATOR;
+    this.css = `.card {
+      font-family: arial;
+      font-size: 20px;
+      text-align: center;
+      color: black;
+    }`;
+
+
```
Comment (ewnd9) [inline src/exporter.js]: Remove unnecessary empty line

File: src/exporter.js [hunk 36]
```diff
@@ -0,0 +1,144 @@
...
+    const models = this._getInitialRowValue('col', 'models');
+    const model = getLastItem(models);
+    model.name = this.deckName;
+    model.css = this.css;
+    model.did = this.topDeckId;
+    model.id = topModelId;
+    models[`${topModelId}`] = model;
+    this._update('update col set models=:models where id=1', { ':models': JSON.stringify(models) });
+
+    return this;
```
Comment (ewnd9) [inline src/exporter.js]: Remove this line, it's redundant

File: src/index.js [hunk 37]
```diff
@@ -1,152 +1,29 @@
 'use strict';
 
-let sql = null;
-let template = null;
+import Exporter from './exporter';
+export {SEPARATOR} from './exporter';
 
-if (process.env.APP_ENV === 'browser') {
+const isBrowser = process.env.APP_ENV === 'browser';
```
Comment (ewnd9) [inline src/index.js]: Not a good idea, this statement is supposed to be replaced by `webpack` on bundling step and it's better to point it out explicitly in every `if`

```js
if (process.env.APP_ENV === 'browser') {
```

File: templates/template.sql [hunk 38]
```diff
@@ -1,5 +1,6 @@
 PRAGMA foreign_keys=OFF;
 BEGIN TRANSACTION;
+DROP TABLE IF EXISTS col;
```
Comment (ewnd9) [inline templates/template.sql]: Don't change unrelated stuff, open a separate PR if it's important

File: test/_helpers.js [hunk 39]
```diff
@@ -3,20 +3,20 @@ import path from 'path';
 import Zip from 'jszip';
 import mkdirp from 'mkdirp';
 
-export const addCards = (apkg, list) => list.forEach(({front, back}) => apkg.addCard(front, back));
+export const addCards = (apkg, list) => list.forEach(({ front, back }) => apkg.addCard(front, back));
```
Comment (ewnd9) [inline test/_helpers.js]: Same here, don't change unrelated stuff, leave it for a another separate PR

File: test/test.js [hunk 40]
```diff
@@ -3,23 +3,28 @@ import test from 'ava';
 import 'babel-register';
 import 'babel-polyfill';
 
-import AnkiExport, { SEPARATOR } from '../src/index';
+import AnkiExport, {SEPARATOR} from '../src/index';
```
Comment (ewnd9) [inline test/test.js]: Why you removed white spaces here and added them in `test/_helpers.js` https://github.com/ewnd9/anki-apkg-export/pull/8/files#diff-c193db11b3837e38bd24c06b98cb93ffR6 and at the bottom of this file https://github.com/ewnd9/anki-apkg-export/pull/8/files#diff-c1129c8b045390789fa8ff62f2c6b4a9R68 ?

(white spaces are cool)

File: test/test.js [hunk 41]
```diff
@@ -3,23 +3,28 @@ import test from 'ava';
...
+import sql from 'sql.js';
 
+const template = require('fs').readFileSync(__dirname + '/../templates/template.sql', 'utf-8');
 const tmpDir = '/tmp/';
 const dest = tmpDir + 'result.apkg';
 const destUnpacked = tmpDir + 'unpacked_result';
 const destUnpackedDb = destUnpacked + '/collection.anki2';
 
-test.beforeEach(async () => pify(exec)(`rm -rf ${dest} ${destUnpacked}`));
+test.beforeEach(async() => pify(exec)(`rm -rf ${dest} ${destUnpacked}`));
```
Comment (ewnd9) [inline test/test.js]: Why?

File: templates/template.sql (hunk 38 above)
Comment (ewnd9) [inline templates/template.sql]: Let's try to fix it then instead of editing `SQL`. That changes tell me that, now, somewhere same `db` instance used twice.

Comment (vvscode) [inline templates/template.sql]: Even serial run doesn't help

Comment (ewnd9) [inline templates/template.sql]: See https://github.com/ewnd9/anki-apkg-export/pull/8#discussion_r91520706

I've asked 

```js
export default function(deckName) {
  return new ApkgExporter(deckName, { sql, template });
}
// exporter.js
constructor() {
  ...
  this.db = new sql.Database();
  ...
}
```

Now you have 

```js
const db = new sql.Database();

export default function (deckName) {
  return new Exporter(deckName, {
    db,
    template
  });
}
```

Database is simply never reinitialized

Comment (vvscode) [inline templates/template.sql]: Oh. Thx. I broke my mind ((

File: src/index.js [hunk 42]
```diff
@@ -1,152 +1,28 @@
...
-
-  return {
-    addMedia,
-    addCard,
-    save
-  };
+export default function (deckName) {
+  return new Exporter(deckName, {
+    template,
+    DbClass: sql.Database
```
Comment (ewnd9) [inline src/index.js]: Why do you keep changing my requirements? https://github.com/ewnd9/anki-apkg-export/pull/8#discussion_r91557291

Comment (vvscode) [inline src/index.js]: sorry. I miss. It's  difficult day today

File: src/index.js [hunk 43]
```diff
@@ -1,152 +1,28 @@
 'use strict';
 
-let sql = null;
-let template = null;
+import Exporter from './exporter';
+export { SEPARATOR } from './exporter';
 
-if (process.env.APP_ENV === 'browser') {
+const isBrowser = process.env.APP_ENV === 'browser';
```
Comment (ewnd9) [inline src/index.js]: https://github.com/ewnd9/anki-apkg-export/pull/8#discussion_r91525743

File: test/test.js [hunk 44]
```diff
@@ -35,7 +40,10 @@ test('equals to sample', async t => {
 
 test('check internal structure', async t => {
   // Create deck as in previous example
-  const apkg = new AnkiExport('deck-name');
+  const apkg = new AnkiExport('deck-name', {
+    template,
+    DbClass: sql.Database
```
Comment (ewnd9) [inline test/test.js]: Should be `sql: sql.Database`. 

The reason why this doesn't break tests is you importing `AnkiExport` from `index.js` instead of `exporter.js` https://github.com/ewnd9/anki-apkg-export/pull/8/files#diff-c1129c8b045390789fa8ff62f2c6b4a9R6

Comment (vvscode) [inline test/test.js]: you're right. Can't understand how it make tests pass

Comment (ewnd9) [inline test/test.js]: > Can't understand how it make tests pass

Because `AnkiExport` is (still) exported from `index.js` instead of `exporter.js`. See the second paragraph from my message above https://github.com/ewnd9/anki-apkg-export/pull/8#discussion_r91722957

Comment (vvscode) [inline test/test.js]: invalidate cache via `delete require.cache[require.resolve('./../src/index')]` is good idea?

Comment (vvscode) [inline test/test.js]: I see. But as I understand - `test.js` test hole package. So I just change instantiating exporter module in test

File: test/test.js [hunk 45]
```diff
@@ -3,7 +3,7 @@ import test from 'ava';
 import 'babel-register';
 import 'babel-polyfill';
 
-import AnkiExport, { SEPARATOR } from '../src/index';
+import ankiExport, { SEPARATOR } from '../src/index';
```
Comment (ewnd9) [inline test/test.js]: Revert all the changes in this file. 

I see it was a little ambiguous as using `new` with exported function from `index.js` is unnecessary, but now it's just renaming variables and adding whitespaces which should be a separate concern

File: test/exporter.js [hunk 46]
```diff
@@ -0,0 +1,65 @@
...
+import sinon from 'sinon';
+import proxyquire from 'proxyquire';
+import fs from 'fs';
+import sql from 'sql.js';
+
+import 'babel-register';
+import 'babel-polyfill';
+
+const template = fs.readFileSync(__dirname + '/../templates/template.sql', 'utf-8');
+// Mock Exporter dependencies
```
Comment (ewnd9) [inline test/exporter.js]: Remove this comment

File: test/test.js (hunk 45 above)
Comment (vvscode) [inline test/test.js]: but passing additional params no has not value. Function-generator doesn't process them

Comment (ewnd9) [inline test/test.js]: I was talking about reverting `test/test.js` to the master (https://github.com/ewnd9/anki-apkg-export/blob/master/test/test.js), sorry for the confusion.

Comment (vvscode) [inline test/test.js]: done

**Comment by ewnd9** [comment]
:tada: Thank you for your efforts

**Comment by vvscode** [comment]
sorry for long PR

**Comment by ewnd9** [comment]
It's great you found the time to finish it. Hope the experience with `sinon` and `proxyquire` helps you in future.
